File: app/src/types/assistant.ts
/**
 * AI Data Assistant type contracts.
 *
 * These types define the metadata schema used by the assistant to understand
 * datasets without accessing full raw data.
 *
 * NOTE: This project uses snake_case in frontend types to match backend
 * JSON serialization conventions (consistent with Dataset, Analysis, etc.).
 */

import type { SafeResultSummary } from './resultSummary';

export type ColumnRole =
  | "user_id"
  | "device_id"
  | "session_id"
  | "order_id"
  | "product_id"
  | "event_name"
  | "timestamp"
  | "date"
  | "metric"
  | "dimension"
  | "category"
  | "treatment_group"
  | "label_target"
  | "amount_revenue"
  | "status"
  | "text_field"
  | "unknown";

export type SemanticType =
  | "numeric"
  | "categorical"
  | "datetime"
  | "boolean"
  | "text"
  | "identifier"
  | "unknown";

export type TableType =
  | "user"
  | "order"
  | "event_log"
  | "product"
  | "campaign"
  | "experiment"
  | "transaction"
  | "dimension"
  | "metric_summary"
  | "review_text"
  | "unknown";

export interface ColumnProfile {
  name: string;
  dtype: string;
  semantic_type: SemanticType;
  role: ColumnRole;
  null_count: number;
  null_rate: number;
  unique_count: number;
  unique_rate: number;
  examples: unknown[];
  min?: number | string | null;
  max?: number | string | null;
  mean?: number | null;
  std?: number | null;
  warnings?: string[];
}

export interface TableClassification {
  table_type: TableType;
  confidence: number;
  evidence: string[];
  recommended_analyses: string[];
  warnings?: string[];
}

export interface DatasetProfile {
  dataset_id: string;
  name: string;
  row_count: number;
  column_count: number;
  columns: ColumnProfile[];
  classification: TableClassification;
  quality_warnings: string[];
  generated_at: string;
}

export interface ProfileDatasetRequest {
  dataset_id: string;
  include_examples?: boolean;
}

// ---------------------------------------------------------------------------
// Relationship Inference Types
// ---------------------------------------------------------------------------

export type RelationshipType =
  | "one_to_one"
  | "one_to_many"
  | "many_to_one"
  | "many_to_many"
  | "unknown";

export type RelationshipStatus = "suggested" | "confirmed" | "rejected";

export type RelationshipRiskLevel = "low" | "medium" | "high";

export type RelationshipEvidenceType =
  | "column_name_match"
  | "role_match"
  | "type_compatibility"
  | "uniqueness_signal"
  | "table_type_signal"
  | "value_overlap"
  | "null_rate_check"
  | "manual_confirmation";

export interface RelationshipEvidence {
  type: RelationshipEvidenceType;
  score: number;
  message: string;
}

export interface TableRelationship {
  id: string;
  source_dataset_id: string;
  target_dataset_id: string;
  source_dataset_name: string;
  target_dataset_name: string;
  source_column: string;
  target_column: string;
  relationship_type: RelationshipType;
  confidence: number;
  status: RelationshipStatus;
  evidence: RelationshipEvidence[];
  warnings: string[];
  created_at?: string;
  confirmed_at?: string;
  risk_level?: RelationshipRiskLevel;
  risk_reasons?: string[];
  is_custom?: boolean;
  source?: "inferred" | "manual" | "mixed";
}

export type RelationshipSetNodeRole =
  | "connected"
  | "isolated"
  | "excluded"
  | "reference_only";

export interface RelationshipSetDatasetNode {
  dataset_id: string;
  dataset_name?: string;
  filename?: string;
  role: RelationshipSetNodeRole;
  reason?: string;
  selected_by_user: boolean;
  joinable: boolean;
  included_in_context: boolean;
}

export interface RelationshipSet {
  id: string;
  name: string;
  description?: string;
  /**
   * Dataset nodes selected or retained for this analysis topic.
   * This includes connected nodes and isolated/reference-only nodes.
   */
  dataset_nodes: RelationshipSetDatasetNode[];
  /**
   * Backward-compatible dataset IDs derived from dataset_nodes.
   */
  dataset_ids: string[];
  /**
   * User-confirmed relationship edges.
   */
  relationships: TableRelationship[];
  created_at: string;
  updated_at: string;
  is_default?: boolean;
  source: "inferred" | "manual" | "mixed";
}

export interface RelationshipSetSummary {
  id: string;
  name: string;
  relationship_count: number;
  dataset_count: number;
  updated_at: string;
  is_default?: boolean;
}

export interface InferRelationshipsRequest {
  dataset_ids: string[];
  include_value_overlap?: boolean;
  max_candidates?: number;
}

export interface InferRelationshipsResponse {
  relationships: TableRelationship[];
  generated_at: string;
  warnings: string[];
}

export interface PlanningColumnMetadata {
  name: string;
  dtype?: string;
  semantic_type?: string;
  role?: string;
  null_rate?: number;
  unique_rate?: number;
}

export interface PlanningDatasetMetadata {
  id: string;
  name?: string;
  filename?: string;
  schema: PlanningColumnMetadata[];
  table_type?: string;
  business_category?: string;
  data_type?: string;
  analysis_tags: string[];
  recommended_analyses: string[];
  quality_warnings: string[];
}

export interface PlanningRelationshipMetadata {
  id?: string;
  source_dataset_id: string;
  source_column: string;
  target_dataset_id: string;
  target_column: string;
  relationship_type?: RelationshipType;
  risk_level?: RelationshipRiskLevel;
  status: 'confirmed';
}

export interface BoundedPlanningContext {
  selected_dataset_ids: string[];
  selected_dataset_id?: string;
  datasets: PlanningDatasetMetadata[];
  relationship_set?: {
    id: string;
    name: string;
    dataset_nodes: Array<{
      dataset_id: string;
      dataset_name?: string;
      role: 'connected' | 'isolated' | 'reference_only';
      joinable: boolean;
    }>;
    relationships: PlanningRelationshipMetadata[];
  };
  analysis_history_summary?: SafeResultSummary;
}

// ---------------------------------------------------------------------------
// Analysis Plan Types
// ---------------------------------------------------------------------------

export type RecommendedAnalysisType =
  | "descriptive"
  | "data_overview"
  | "attribution"
  | "forecast"
  | "path_analysis"
  | "ab_test"
  | "regression"
  | "smart_process"
  | "visualization";

export type AnalysisPlanSource = "hermes_live" | "deterministic_fallback";

export type AnalysisPlanConfidence = "low" | "medium" | "high";

export type AnalysisExecutionReadiness =
  | "ready_single_table"
  | "needs_join"
  | "needs_clarification"
  | "unsupported";

export type AnalysisPlanNextAction =
  | "review_plan"
  | "navigate_analysis"
  | "create_analysis_dataset"
  | "clarify";

export interface AnalysisFieldRequirement {
  dataset_id: string;
  role:
    | "target_metric"
    | "time_column"
    | "user_id"
    | "group_column"
    | "event_name"
    | "dimension"
    | "feature"
    | "join_key";
  required: boolean;
  candidate_columns: string[];
  reason: string;
}

export interface AssistantNextAction {
  type: "navigate" | "confirm" | "explain" | "warning";
  label: string;
  target?: string;
  payload?: Record<string, unknown>;
}

export interface AssistantCandidateDataset {
  dataset_id: string;
  dataset_name?: string;
  reasons: string[];
  confidence: "low" | "medium" | "high";
}

export interface AnalysisRelationshipRequirement {
  relationship_id?: string;
  source_dataset_id: string;
  source_column: string;
  target_dataset_id: string;
  target_column: string;
  status: "confirmed" | "requires_confirmation";
  relationship_type?: RelationshipType;
  risk_level?: RelationshipRiskLevel;
  reason: string;
}

export interface AnalysisMetricTarget {
  name: string;
  dataset_id: string;
  field?: string;
  aggregation?: "count" | "count_distinct" | "sum" | "average" | "min" | "max" | "rate";
  description: string;
}

export interface AssistantAnalysisPlan {
  id: string;
  user_question: string;
  interpreted_goal: string;
  recommended_analysis_type: RecommendedAnalysisType;
  required_datasets: string[];
  required_dataset_ids: string[];
  candidate_dataset_ids: string[];
  candidate_datasets: AssistantCandidateDataset[];
  required_fields: AnalysisFieldRequirement[];
  required_relationships: AnalysisRelationshipRequirement[];
  metrics: AnalysisMetricTarget[];
  relationship_set_id?: string;
  relationship_set_name?: string;
  reference_dataset_ids: string[];
  assumptions: string[];
  warnings: string[];
  clarifying_questions: string[];
  execution_readiness: AnalysisExecutionReadiness;
  next_action: AnalysisPlanNextAction;
  next_actions: AssistantNextAction[];
  source: AnalysisPlanSource;
  confidence: AnalysisPlanConfidence;
  fallback_used: boolean;
}

export interface AnalysisPrefillPayload {
  source: "ai_workbench";
  plan_id: string;
  analysis_type: RecommendedAnalysisType;
  dataset_ids: string[];
  primary_dataset_id?: string;
  relationship_set_id?: string;
  relationship_set_name?: string;
  suggested_fields?: AnalysisFieldRequirement[];
  user_question?: string;
  created_at: string;
}

File: insightease-backend/app/schemas/assistant.py
"""Canonical schemas for metadata-first Assistant capabilities."""

from typing import Annotated, Any, Dict, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


class StrictAssistantModel(BaseModel):
    """Reject fields outside the documented metadata/plan contract."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)


class RelationshipEvidence(StrictAssistantModel):
    type: str
    score: float
    message: str


class TableRelationship(StrictAssistantModel):
    id: str
    source_dataset_id: str
    target_dataset_id: str
    source_dataset_name: str
    target_dataset_name: str
    source_column: str
    target_column: str
    relationship_type: str
    confidence: float
    status: str
    evidence: List[RelationshipEvidence]
    warnings: List[str]
    created_at: Optional[str] = None
    confirmed_at: Optional[str] = None


class InferRelationshipsRequest(StrictAssistantModel):
    dataset_ids: List[str]
    include_value_overlap: bool = False
    max_candidates: int = Field(default=200, ge=1, le=1000)


class InferRelationshipsResponse(StrictAssistantModel):
    relationships: List[TableRelationship]
    generated_at: str
    warnings: List[str]


# ---------------------------------------------------------------------------
# Bounded planning context
# ---------------------------------------------------------------------------

PlanSource = Literal["hermes_live", "deterministic_fallback"]
PlanConfidence = Literal["low", "medium", "high"]
ExecutionReadiness = Literal[
    "ready_single_table",
    "needs_join",
    "needs_clarification",
    "unsupported",
]
PlanNextAction = Literal[
    "review_plan",
    "navigate_analysis",
    "create_analysis_dataset",
    "clarify",
]
RecommendedAnalysisType = Literal[
    "descriptive",
    "data_overview",
    "attribution",
    "forecast",
    "path_analysis",
    "ab_test",
    "regression",
    "smart_process",
    "visualization",
]
AnalysisFieldRole = Literal[
    "target_metric",
    "time_column",
    "user_id",
    "group_column",
    "event_name",
    "dimension",
    "feature",
    "join_key",
]
BoundedPlanText = Annotated[str, Field(min_length=1, max_length=500)]


class PlanningColumnMetadata(StrictAssistantModel):
    name: str = Field(..., min_length=1, max_length=200)
    dtype: Optional[str] = Field(default=None, max_length=100)
    semantic_type: Optional[str] = Field(default=None, max_length=100)
    role: Optional[str] = Field(default=None, max_length=100)
    null_rate: Optional[float] = Field(default=None, ge=0, le=1)
    unique_rate: Optional[float] = Field(default=None, ge=0, le=1)


class PlanningDatasetMetadata(StrictAssistantModel):
    id: str = Field(..., min_length=1, max_length=200)
    name: Optional[str] = Field(default=None, max_length=255)
    filename: Optional[str] = Field(default=None, max_length=255)
    columns: List[PlanningColumnMetadata] = Field(
        default_factory=list,
        alias="schema",
        serialization_alias="schema",
        max_length=100,
    )
    table_type: Optional[str] = Field(default=None, max_length=100)
    business_category: Optional[str] = Field(default=None, max_length=100)
    data_type: Optional[str] = Field(default=None, max_length=100)
    analysis_tags: List[str] = Field(default_factory=list, max_length=20)
    recommended_analyses: List[str] = Field(default_factory=list, max_length=20)
    quality_warnings: List[str] = Field(default_factory=list, max_length=20)


class PlanningRelationshipSetNode(StrictAssistantModel):
    dataset_id: str = Field(..., min_length=1, max_length=200)
    dataset_name: Optional[str] = Field(default=None, max_length=255)
    role: Literal["connected", "isolated", "reference_only"]
    joinable: bool


class PlanningRelationshipMetadata(StrictAssistantModel):
    id: Optional[str] = Field(default=None, max_length=500)
    source_dataset_id: str = Field(..., min_length=1, max_length=200)
    source_column: str = Field(..., min_length=1, max_length=200)
    target_dataset_id: str = Field(..., min_length=1, max_length=200)
    target_column: str = Field(..., min_length=1, max_length=200)
    relationship_type: Optional[Literal[
        "one_to_one",
        "one_to_many",
        "many_to_one",
        "many_to_many",
        "unknown",
    ]] = None
    risk_level: Optional[Literal["low", "medium", "high"]] = None
    status: Literal["confirmed"] = "confirmed"


class PlanningRelationshipSet(StrictAssistantModel):
    id: str = Field(..., min_length=1, max_length=200)
    name: str = Field(..., min_length=1, max_length=255)
    dataset_nodes: List[PlanningRelationshipSetNode] = Field(default_factory=list, max_length=50)
    relationships: List[PlanningRelationshipMetadata] = Field(default_factory=list, max_length=200)


class BoundedPlanningContext(StrictAssistantModel):
    selected_dataset_ids: List[str] = Field(default_factory=list, max_length=20)
    selected_dataset_id: Optional[str] = Field(default=None, max_length=200)
    datasets: List[PlanningDatasetMetadata] = Field(default_factory=list, max_length=20)
    relationship_set: Optional[PlanningRelationshipSet] = None
    analysis_history_summary: Optional[Dict[str, Any]] = None


# ---------------------------------------------------------------------------
# Canonical advisory AnalysisPlan
# ---------------------------------------------------------------------------


class AnalysisFieldRequirement(StrictAssistantModel):
    dataset_id: str = Field(..., min_length=1, max_length=200)
    role: AnalysisFieldRole
    required: bool
    candidate_columns: List[str] = Field(default_factory=list, max_length=20)
    reason: str = Field(..., min_length=1, max_length=500)


class AnalysisRelationshipRequirement(StrictAssistantModel):
    relationship_id: Optional[str] = Field(default=None, max_length=500)
    source_dataset_id: str = Field(..., min_length=1, max_length=200)
    source_column: str = Field(..., min_length=1, max_length=200)
    target_dataset_id: str = Field(..., min_length=1, max_length=200)
    target_column: str = Field(..., min_length=1, max_length=200)
    status: Literal["confirmed", "requires_confirmation"]
    relationship_type: Optional[Literal[
        "one_to_one",
        "one_to_many",
        "many_to_one",
        "many_to_many",
        "unknown",
    ]] = None
    risk_level: Optional[Literal["low", "medium", "high"]] = None
    reason: str = Field(..., min_length=1, max_length=500)


class AnalysisMetricTarget(StrictAssistantModel):
    name: str = Field(..., min_length=1, max_length=200)
    dataset_id: str = Field(..., min_length=1, max_length=200)
    field: Optional[str] = Field(default=None, max_length=200)
    aggregation: Optional[Literal["count", "count_distinct", "sum", "average", "min", "max", "rate"]] = None
    description: str = Field(..., min_length=1, max_length=500)


class AssistantCandidateDataset(StrictAssistantModel):
    dataset_id: str = Field(..., min_length=1, max_length=200)
    dataset_name: Optional[str] = Field(default=None, max_length=255)
    reasons: List[BoundedPlanText] = Field(default_factory=list, max_length=8)
    confidence: PlanConfidence


class AssistantNextAction(StrictAssistantModel):
    type: Literal["navigate", "confirm", "explain", "warning"]
    label: str = Field(..., min_length=1, max_length=200)
    target: Optional[str] = Field(default=None, max_length=300)
    payload: Optional[Dict[str, Any]] = None


class AssistantAnalysisPlan(StrictAssistantModel):
    id: str = Field(..., min_length=1, max_length=200)
    user_question: str = Field(..., min_length=1, max_length=2000)
    interpreted_goal: str = Field(..., min_length=1, max_length=1000)
    recommended_analysis_type: RecommendedAnalysisType
    required_datasets: List[str] = Field(default_factory=list, max_length=20)
    required_dataset_ids: List[str] = Field(default_factory=list, max_length=20)
    candidate_dataset_ids: List[str] = Field(default_factory=list, max_length=20)
    candidate_datasets: List[AssistantCandidateDataset] = Field(default_factory=list, max_length=20)
    required_fields: List[AnalysisFieldRequirement] = Field(default_factory=list, max_length=100)
    required_relationships: List[AnalysisRelationshipRequirement] = Field(default_factory=list, max_length=50)
    metrics: List[AnalysisMetricTarget] = Field(default_factory=list, max_length=20)
    relationship_set_id: Optional[str] = Field(default=None, max_length=200)
    relationship_set_name: Optional[str] = Field(default=None, max_length=255)
    reference_dataset_ids: List[str] = Field(default_factory=list, max_length=20)
    assumptions: List[BoundedPlanText] = Field(default_factory=list, max_length=20)
    warnings: List[BoundedPlanText] = Field(default_factory=list, max_length=20)
    clarifying_questions: List[BoundedPlanText] = Field(default_factory=list, max_length=8)
    execution_readiness: ExecutionReadiness
    next_action: PlanNextAction
    next_actions: List[AssistantNextAction] = Field(default_factory=list, max_length=8)
    source: PlanSource
    confidence: PlanConfidence
    fallback_used: bool
